### PR TITLE
Dataloader explained and maintanance categorical

### PR DESCRIPTION
**Pull request by tiaguinho-code**
# Dataloader tutorial 
Added a cool little jupyter notebook showcasing the user features of the dataloader

# Resuscitation of categorical preprocessor maintainance 
I wasn't really taking too much care of the categorical thing as I thought you guys wanted the one hot / binarized one. Anyway, started fixing some stuff here too.

### CODE REVIEW

File: tutorials/how_to_dataloader.ipynb [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+      ]
+     },
+     "execution_count": 8,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "from sklearn.metrics import confusion_matrix\n",
+    "y_true_cm = y_pred_cm = data_loader.confusion_matrix_data_adapter_categorical(y_test)\n",
```
Comment (mlederbauer) [inline tutorials/how_to_dataloader.ipynb:300]: Nice! Suggestion to expand this: For our final submission, we could also show how the confusion matrix changes for i) a "dummy model" that only chooses the most often occuring ligand and ii) a "simple" Random Forest model that we use in our workflow, but without hyperparameter optimization (that perhaps shows a better performance than the dummy model)

File: tutorials/how_to_dataloader.ipynb [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "[[-0.2722084  -2.21144124 -1.43206187  3.74585072 -5.16278678 -3.93518974\n",
+      "   0.          0.          1.          0.        ]\n",
```
Comment (mlederbauer) [inline tutorials/how_to_dataloader.ipynb:140]: Where do those numbers come from (after the NMR tensors) ? Are those the metal, E, X2, X2, X3 ligands? This is quite hard for me to understand atm

File: tutorials/how_to_dataloader.ipynb [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+   "source": [
+    "print(y_test)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Decode target (y) arrays\n",
+    "To decode the targets you got you can use ether the `categorical_target_decoder()` or the `binarized_target_decoder()` method of the dataloader. This will return a labled list of lists. As far as I know these decoded things can be easily passed into the cm or maybe after flattening them."
```
Comment (mlederbauer) [inline tutorials/how_to_dataloader.ipynb]: What is a labeled list of lists? Not entirely sure what that means 😅 Would a mapping in a dictionary be better ?

File: tutorials/how_to_dataloader.ipynb [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "[[ 0 10 13]\n",
```
Comment (mlederbauer) [inline tutorials/how_to_dataloader.ipynb:175]: so here column 1 = metal, col 2 = E, col 3=X4?

File: nmrcraft/data/dataset.py [hunk 5]
```diff
@@ -278,6 +278,35 @@ def more_than_one_target(self):
...
+                target_encoders[i].inverse_transform(target_column)
+            )
+
+        # Rotate back so each row corresponds to a complex and not the target like metal or X4
+        ys_decoded_properly_rotated = [
+            list(x) if i == 0 else x
+            for i, x in enumerate(map(list, zip(*ys_decoded)))
+        ]
+
+        return ys_decoded_properly_rotated
```
Comment (mlederbauer) [inline nmrcraft/data/dataset.py]: Nice, thank you for the addition and testing it with the scikit learn confusion matrix, that's super valuable! I think the code needs more commenting to make 100% clear what each function/part does.

Comment (mlederbauer) [reply nmrcraft/data/dataset.py]: e.g., what does decode, rotate, binarize, ... mean. We can add this as docstrings with examples to the function definition

**Review by mlederbauer — APPROVED** [review]
Many thanks for this addition! Could you kindly address the clarification questions I had for the dataloader ? Some functionality is not yet clear to me

File: tutorials/how_to_dataloader.ipynb (hunk 2 above)
Comment (tiaguinho-code) [reply tutorials/how_to_dataloader.ipynb:140]: Yeah, I'll explain that

File: tutorials/how_to_dataloader.ipynb (hunk 4 above)
Comment (tiaguinho-code) [reply tutorials/how_to_dataloader.ipynb:175]: exactly! , I will add this to the markdown part

File: nmrcraft/data/dataset.py (hunk 5 above)
Comment (tiaguinho-code) [reply nmrcraft/data/dataset.py]: I will take care of the whole rotatation stuff when refactoring. I already have a decent idea of how to take the code apart to make it clearer. But I definetly agree that "rotate" makes it hard to understand 😆

File: tutorials/how_to_dataloader.ipynb (hunk 3 above)
Comment (tiaguinho-code) [reply tutorials/how_to_dataloader.ipynb]: It's a list of lists (basically an array) that contains the Label in each position instead of a meaningless number. So "Mo" or "Ethyl" instead of 1 or 7

**Comment by tiaguinho-code** [comment]
Alright, did some refactoring and also some additions to docs. Let me know if anything still needs to be clearer but in principle this is ready to merge so the models can be worked on further.

**Comment by mlederbauer** [comment]
Thank you! Merging it

File: tutorials/how_to_dataloader.ipynb (hunk 3 above)
Comment (mlederbauer) [reply tutorials/how_to_dataloader.ipynb]: Super, thanks! Maybe good to mention this in the docs with explicit examples
